pkg/controller/direct/bigquerydatatransfer/bigquerydatatransferconfig_controller.go

```diff
@@ -0,0 +1,351 @@
+}
+
+func NewModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &model{config: *config}, nil
+}
+
+var _ directbase.Model = &model{}
+
+type model struct {
+	config config.ControllerConfig
+}
+
+func (m *model) client(ctx context.Context) (*gcp.Client, error) {
+	var opts []option.ClientOption
+	if m.config.UserAgent != "" {
+		opts = append(opts, option.WithUserAgent(m.config.UserAgent))
+	}
+	if m.config.HTTPClient != nil {
+		opts = append(opts, option.WithHTTPClient(m.config.HTTPClient))
+	}
+	if m.config.UserProjectOverride && m.config.BillingProject != "" {
+		opts = append(opts, option.WithQuotaProject(m.config.BillingProject))
+	}
+
+	gcpClient, err := gcp.NewRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building bigquerydatatransfer client: %w", err)
+	}
+	return gcpClient, err
+}
+
+func (m *model) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.BigQueryDataTransferConfig{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	// Get ResourceID
+	resourceID := direct.ValueOf(obj.Spec.ResourceID)
+	if resourceID == "" {
+		serviceGeneratedID, err := parseServiceGeneratedIDFromExternalRef(obj)
+		if err != nil {
+			return nil, err
+		}
+		resourceID = serviceGeneratedID
+	}
+	// TODO(jingyih): need to figure out how to deal with resourceID. It is service generated.
+
+	location := obj.Spec.Location
+	if location == "" {
+		return nil, fmt.Errorf("cannot resolve location")
+	}
+
+	// Resolve PubSubTopic Ref
+	if obj.Spec.PubSubTopicRef != nil {
+		topic, err := refv1beta1.ResolvePubSubTopic(ctx, reader, obj, obj.Spec.PubSubTopicRef)
+		if err != nil {
+			return nil, err
+		}
+		obj.Spec.PubSubTopicRef.External = topic.String()
+	}
+
+	// Resolve BigQueryDataSet Ref
+	if obj.Spec.DatasetRef != nil {
+		dataset, err := refv1beta1.ResolveBigQueryDataset(ctx, reader, obj, obj.Spec.DatasetRef)
+		if err != nil {
+			return nil, err
+		}
+		obj.Spec.DatasetRef.External = dataset.String()
+	}
+
+	// Resolve KMSCryptoKey Ref
+	if obj.Spec.EncryptionConfiguration != nil {
+		kmsCryptoKeyRef, err := refv1beta1.ResolveKMSCryptoKeyRef(ctx, reader, obj, obj.Spec.EncryptionConfiguration.KmsKeyRef)
+		if err != nil {
+			return nil, err
+		}
+		obj.Spec.EncryptionConfiguration.KmsKeyRef = kmsCryptoKeyRef
+	}
+
+	// Resolve Project Ref
+	projectRef, err := refv1beta1.ResolveProject(ctx, reader, obj, obj.Spec.ProjectRef)
+	if err != nil {
+		return nil, err
+	}
+	projectID := projectRef.ProjectID
+	if projectID == "" {
+		return nil, fmt.Errorf("cannot resolve project")
+	}
+
+	var id *BigQueryDataTransferConfigIdentity
+	externalRef := direct.ValueOf(obj.Status.ExternalRef)
+	if externalRef == "" {
+		id = BuildID(projectID, location, resourceID)
```

Comment:
So the problem here is that since we do not assign `metadata.name` to resourceID, if `spec.resourceID` is not given and the GCP object is not created (no `status.externalRef`), this will always pass in an empty `resourceID`, which fails the following logics.